src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <regex>
+#include <vector>
+
+#include "generation_config_helper.hpp"
+#include "openvino/genai/llm_pipeline.hpp"
+#include "utils.hpp"
+
+
+namespace {
+
+struct TokenIdScore {
+    int id;
+    float score;
+
+    TokenIdScore() = default;
+    TokenIdScore(int id, float score) : id(id), score(score) {}
+
+    bool operator<(const TokenIdScore& other) const {
+        return score < other.score;
+    }
+    bool operator>(const TokenIdScore& other) const {
+        return score > other.score;
+    }
+};
+
+void sampling_softmax_inplace(TokenIdScore* first, TokenIdScore* last) {

```

Comment:
You always call it `sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());` wit the full range. Take a ref to the vector instad

Reply:
done